### PR TITLE
Enforce use of throwing `AbsolutePath` initializer

### PR DESCRIPTION
**Pull request by neonichu**
In support of https://github.com/apple/swift-package-manager/pull/5797

### CODE REVIEW

File: Sources/TSCTestSupport/FileSystemExtensions.swift [hunk 1]
```diff
@@ -72,3 +72,19 @@ extension FileSystem {
         }
     }
 }
+
+public extension AbsolutePath {
+    init(path: StaticString) {
```
Comment (neonichu) [inline Sources/TSCTestSupport/FileSystemExtensions.swift:77]: Without an argument label, the `String` version of the initializer seemed to always get chosen, so I ended up just adding an argument label. These initializers are a convenience for tests which often deal with static paths.

**Comment by neonichu** [comment]
@swift-ci please test

File: Sources/TSCUtility/PkgConfig.swift [hunk 2]
```diff
@@ -47,12 +47,14 @@ public struct PCFileFinder {
...
     /// By default, this is combined with the search paths inferred from
     /// `pkg-config` itself.
-    static let searchPaths = [
-        AbsolutePath("/usr/local/lib/pkgconfig"),
-        AbsolutePath("/usr/local/share/pkgconfig"),
-        AbsolutePath("/usr/lib/pkgconfig"),
-        AbsolutePath("/usr/share/pkgconfig"),
-    ]
+    static func searchPaths() throws -> [AbsolutePath] {
+        return try [
```
Comment (tomerd) [inline Sources/TSCUtility/PkgConfig.swift:51]: does this not use the non-throwing static init?

File: Sources/TSCUtility/Platform.swift [hunk 3]
```diff
@@ -49,19 +49,21 @@ public enum Platform: Equatable {
...
-            return .android
-        }
-        if fileSystem.isFile(AbsolutePath("/etc/redhat-release")) ||
-            fileSystem.isFile(AbsolutePath("/etc/centos-release")) ||
-            fileSystem.isFile(AbsolutePath("/etc/fedora-release")) ||
-            Platform.isAmazonLinux2(fileSystem) {
-            return .linux(.fedora)
-        }
+        do {
+            if try fileSystem.isFile(AbsolutePath("/etc/debian_version")) {
```
Comment (tomerd) [inline Sources/TSCUtility/Platform.swift:53]: should we have a non-throwing initializer for static string?

**Review by tomerd — APPROVED** [review]
looks great. couple of small questions

File: Sources/TSCUtility/Platform.swift (hunk 3 above)
Comment (neonichu) [reply Sources/TSCUtility/Platform.swift:53]: I'd like to keep that to tests since it is pretty rare for us to construct paths from static strings.

**Comment by neonichu** [comment]
```
/home/build-user/swift-tools-support-core/Sources/TSCBasic/FileSystem.swift:358:33: error: extraneous argument label 'validating:' in call
        return try? AbsolutePath(validating: String(cString: fsr))
                                ^~~~~~~~~~~~~
```

Interesting, not sure why this didn't show up locally

**Comment by neonichu** [comment]
@swift-ci please test

**Comment by neonichu** [comment]
@swift-ci please test

**Comment by neonichu** [comment]
@swift-ci please test

**Comment by neonichu** [comment]
Looks like at least on macOS we're reaching the swift-driver build now.

**Comment by neonichu** [comment]
@swift-ci please test

**Comment by neonichu** [comment]
Updated the approach slightly here, I am keeping `AbsolutePath(validating:)` and changed the relative path one to `AbsolutePath(validating:relativeTo:)`. These changes, plus an extension that offers the non-throwing variants as deprecated, allow clients to move to the throwing variants piecemeal. 

In a future PR, we can remove the deprecated initializers and `validating` argument labels.

**Comment by neonichu** [comment]
@swift-ci please test

**Comment by neonichu** [comment]
https://github.com/apple/swift-driver/pull/1204
https://github.com/apple/swift-tools-support-core/pull/353
https://github.com/apple/sourcekit-lsp/pull/655
@swift-ci please test

**Comment by neonichu** [comment]
https://github.com/apple/swift-package-manager/pull/5797
https://github.com/apple/swift-driver/pull/1204
https://github.com/apple/sourcekit-lsp/pull/655
@swift-ci please test

File: Sources/TSCBasic/FileSystem.swift [hunk 4]
```diff
@@ -353,7 +355,7 @@ private class LocalFileSystem: FileSystem {
         let fsr: UnsafePointer<Int8> = cwdStr.fileSystemRepresentation
         defer { fsr.deallocate() }
 
-        return try? AbsolutePath(validating: String(cString: fsr))
```
Comment (WowbaggersLiquidLunch) [inline Sources/TSCBasic/FileSystem.swift:356]: It seems to me this broke SwiftPM.

**Pull request merged**